### PR TITLE
Gb/multi res

### PR DESCRIPTION
**Pull request by grantbuster**
Multi-resolution resource handling feature. I want this to be tested on Eagle with dev installs before we merge.

### CODE REVIEW

File: rex/renewable_resource.py [hunk 1]
```diff
@@ -500,15 +510,16 @@ def _get_calculated_ds(self, val, ds_name, var_name, ds_slice):
         out = linear_interp(ts1, v1, ts2, v2, val)
         return out
 
-    def _set_sam_res(self, values, dsets, SAM_res, time_slice, sites):
+    @staticmethod
+    def _set_sam_res(res, values, dsets, SAM_res, time_slice, sites):
```
Comment (bnb32) [inline rex/renewable_resource.py:514]: now that this is being touched again maybe we should use the opportunity to add to the doc string?

File: tests/test_multi_res_resource.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            lr_data = lr_res[dset, :, lr_gids]
+            hr_data = mrr[dset]
+            assert np.allclose(lr_data, hr_data[::12])
+
+            for gid in gids_hr:
+                lr_gids = tree.query(hr_res.coordinates[gid])[1]
+                lr_data = lr_res[dset, :, lr_gids]
+                hr_data = mrr[dset, :, gid]
+                assert len(lr_data.shape) == len(hr_data.shape)
+                assert np.allclose(lr_data, hr_data[::12])
```
Comment (bnb32) [inline tests/test_multi_res_resource.py:105]: can we define this step size as a global var at the top. didn't immediately know where this 12 came from.

File: tests/test_sam_resource.py [hunk 3]
```diff
@@ -246,6 +245,16 @@ def test_preload_sam_sites(sites):
     assert np.allclose(truth, test)
 
 
+def test_bad_site_req():
+    """Test that the preload_SAM method raises an error on a bad site request
+    """
+    sites = [0, 10, 500]
```
Comment (bnb32) [inline tests/test_sam_resource.py:251]: why are these bad sites? it is bc there arent 500 sites in the h5 here?

**Review by bnb32 — CHANGES_REQUESTED** [review]

File: rex/renewable_resource.py (hunk 1 above)
Comment (grantbuster) [reply rex/renewable_resource.py:514]: yep good call!

File: tests/test_multi_res_resource.py (hunk 2 above)
Comment (grantbuster) [reply tests/test_multi_res_resource.py:105]: can do

File: tests/test_sam_resource.py (hunk 3 above)
Comment (grantbuster) [reply tests/test_sam_resource.py:251]: correct, will add comment

File: rex/multi_res_resource.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                    out = copy.deepcopy(lr_attr)
+                    out.update(hr_attr)
+                    return out
+            except Exception as e:
+                msg = ('Could not retrieve attribute "{}" from '
+                       'MultiResolutionResource handler, the hr and lr '
+                       'handler attributes could not be combined: {} {}'
+                       .format(attr, hr_attr, lr_attr))
+                logger.error(msg)
+                raise RuntimeError(msg) from e
```
Comment (ppinchuk) [inline rex/multi_res_resource.py:274]: Any particular reason for `RuntimeError` here? I think generally `__getattr__` raises `AttributeError`

File: rex/multi_res_resource.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        else:
+            try:
+                hr_attr = getattr(self._hr_res, attr)
+                lr_attr = getattr(self._lr_res, attr)
+                if hasattr(hr_attr, '__call__'):
+                    return hr_attr
+                elif isinstance(hr_attr, list) and isinstance(lr_attr, list):
+                    return list(set(hr_attr + lr_attr))
+                elif isinstance(hr_attr, tuple) and isinstance(lr_attr, tuple):
+                    return hr_attr + lr_attr
```
Comment (ppinchuk) [inline rex/multi_res_resource.py:263]: Does this not need the `set(hr_attr + lr_attr)` treatment that list values got above?

File: rex/multi_res_resource.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+        cls_kwargs = dict(nn_map=nn_map, nn_d=nn_d,
+                          handler_class=handler_class,
+                          handle_kwargs=handle_kwargs)
+
+        with cls(h5_hr, h5_lr, **cls_kwargs) as mrr:
+            sig = signature(mrr._hr_res._preload_SAM)
+            preload_kwargs = {k: v for k, v in kwargs.items()
+                              if k in sig.parameters}
+            SAM_res = mrr._hr_res._preload_SAM(mrr, sites, *args,
```
Comment (ppinchuk) [inline rex/multi_res_resource.py:337]: Yea, makes sense why you had to update `_preload_SAM` to a static method

**Review by ppinchuk — DISMISSED** [review]
Nice work. The dynamic attributes can definitely get complicated, but I really like how much flexibility they give us with some of these handlers, so I am a fan.

All my comments are take it or leave it

File: rex/multi_res_resource.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+            gid's are based on this dataset, and the lr_res data is mapped to
+            this.
+        lr_res : Resource | MultiFileResource | MultiYearResource
+            rex resource handler for the low-resolution data. The data from
+            this handler is mapped to the hr_res data.
+
+        Returns
+        -------
+        nn_d : np.ndarray
+            Optional 1D array of nearest neighbor distances. This will be
```
[the remark this answers is not in the document]
Comment (grantbuster) [reply rex/multi_res_resource.py]: Yeah sorry! i just copied the original docstring. Removed the unnecessary text.

File: rex/multi_res_resource.py (hunk 4 above)
Comment (grantbuster) [reply rex/multi_res_resource.py]: Agree, forgot about attrerror, I usually just default to runtimeerror haha

File: rex/multi_res_resource.py (hunk 5 above)
Comment (grantbuster) [reply rex/multi_res_resource.py]: I could have sworn there was a tuple that should have just gotten summed but i cant find it so i yes i agree, tuple(set())

**Pull request merged**